db/sg_replicate_util.go

```diff
@@ -12,36 +12,70 @@ package db
 
 import (
 	"errors"
+	"fmt"
 	"net/http"
 
 	"github.com/couchbase/sync_gateway/base"
 )
 
-// QueryParams retrieves the channels associated with the byChannels a replication filter
-// from the generic queryParams interface{}.
-// The Channels may be passed as a JSON array of strings directly,
-// or embedded in a JSON object with the "channels" property and array value
-func ChannelsFromQueryParams(queryParams interface{}) (channels []string, err error) {
+// CollectionChannelsFromQueryParams returns the channels associated with the byChannel replication filter from the generic queryParams.
+// The channels may be passed in one of two ways:
+//
+//  1. As a JSON array of strings directly:
+//     ["channel1", "channel2"]
+//
+//  2. As a JSON array of strings embedded in a JSON object with the "channels" property:
+//     {"channels": ["channel1", "channel2"] }
+//
+// 3. As a JSON object with a property for each collection, with the value being an array of channels:
+//
+//	{
+//	  "collections_channels": {
+//	    "collection1": ["scope1.channel1"],
+//	    "collection2": ["scope1.channel1", "scope1.channel2"]
+//	  }
+//	}
+func CollectionChannelsFromQueryParams(namedCollections []string, queryParams interface{}) (perCollectionChannels [][]string, allCollectionsChannels []string, err error) {
+	switch val := queryParams.(type) {
+	case map[string]interface{}:
+		if chanarray, ok := val["channels"].([]interface{}); ok {
+			allCollectionsChannels, err = interfaceValsToStringVals(chanarray)
+			return nil, allCollectionsChannels, err
+		}
 
-	var chanarray []interface{}
-	if paramsmap, ok := queryParams.(map[string]interface{}); ok {
-		if chanarray, ok = paramsmap["channels"].([]interface{}); !ok {
-			return nil, errors.New("Replication specifies sync_gateway/bychannel filter, but query_params is missing channels property")
+		if collectionChannels, ok := val["collections_channels"].(map[string]interface{}); ok {
```

Comment:
You could also check for both parameters here.

In order of least surprise to me would be the following behaviors. As long as we document and test the behavior I could be persuaded to use any of them.

1. `channels` is only for default collection and `collection_channels` is only for multiple collections, like the principal documents.
2. `channels` and `collection_channels` are mutually exclusive, but `channels` will imply collection1.channel1, collection2.channel1...
3. `channels` overrides `collection_channels` silently (what I think is the current behavior?)
4. `collection_channels` overrides channels silently
5. `collection_channels` takes precedence, but there's fallback behavior to channels if specified.

Reply:
Decided to go with 2 - I can see it being useful.